x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -951,4 +967,82 @@ protected String contentType() {
     public FieldMapper.Builder getMergeBuilder() {
         return new Builder(simpleName(), indexVersionCreated).init(this);
     }
+
+    @Override
+    public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
+        if (copyTo.copyToFields().isEmpty() != true) {
+            throw new IllegalArgumentException(
+                "field [" + name() + "] of type [" + typeName() + "] doesn't support synthetic source because it declares copy_to"
+            );
+        }
+        return new WildcardSyntheticFieldLoader();
+    }
+
+    private class WildcardSyntheticFieldLoader implements SourceLoader.SyntheticFieldLoader {
+        private final ByteArrayStreamInput docValuesStream = new ByteArrayStreamInput();
+        private int docValueCount;
+        private BytesRef docValueBytes;
+
+        private List<Object> storedValues = emptyList();
+
+        @Override
+        public Stream<Map.Entry<String, StoredFieldLoader>> storedFieldLoaders() {
+            if (ignoreAbove != Defaults.IGNORE_ABOVE) {
+                return Stream.of(Map.entry(originalName(), storedValues -> this.storedValues = storedValues));
+            }
+            return Stream.empty();
+        }
+
+        @Override
+        public DocValuesLoader docValuesLoader(LeafReader leafReader, int[] docIdsInLeaf) throws IOException {
+            BinaryDocValues values = leafReader.getBinaryDocValues(name());
+            if (values == null) {
+                docValueCount = 0;
+                return null;
+            }
+
+            return docId -> {
```

Comment:
Double checking, if a document has one or more values for a wildcard field then we also create a doc values field? So there is no validation needed in case doc values has been disabled? (I don't think this is possible, from looking at `WildcaedFieldMapper` and it always seem to use `CustomBinaryDocValuesField`)